core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala

```diff
@@ -0,0 +1,97 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package integration.kafka.server
+
+import kafka.server.{BaseFetchRequestTest, BrokerTopicStats, KafkaConfig}
+import kafka.utils.{TestInfoUtils, TestUtils}
+import org.apache.kafka.clients.admin.Admin
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.protocol.ApiKeys
+import org.junit.jupiter.api.{AfterEach, BeforeEach, Timeout}
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
+
+import java.util.Properties
+
+class FetchFromFollowerIntegrationTest extends BaseFetchRequestTest {
+  val numNodes = 2
+  val numParts = 1
+  val initialMessages = 100
+  val nMessages = 100
+
+  val topic = "test-fetch-from-follower"
+  val leaderBrokerId = 0
+  val followerBrokerId = 1
+  var admin: Admin = null
+
+  def overridingProps: Properties = {
+    val props = new Properties
+    props.put(KafkaConfig.NumPartitionsProp, numParts.toString)
+
+    props
+  }
+
+  override def generateConfigs: collection.Seq[KafkaConfig] =
+    TestUtils.createBrokerConfigs(numNodes, zkConnectOrNull, enableControlledShutdown = false, enableFetchFromFollower = true)
+      .map(KafkaConfig.fromProps(_, overridingProps))
+
+  @BeforeEach
+  def initializeFetchFromFollowerCluster(): Unit = {
+    // Create a 2 broker cluster where broker 0 is the leader and 1 is the follower.
+
+    admin = TestUtils.createAdminClient(brokers, listenerName)
+    TestUtils.createTopicWithAdminRaw(
+      admin,
+      topic,
+      replicaAssignment = Map(0 -> Seq(leaderBrokerId, followerBrokerId))
+    )
+    TestUtils.generateAndProduceMessages(brokers, topic, initialMessages)
+  }
+
+  @AfterEach
+  def close(): Unit = {
+    if (admin != null) {
+      admin.close()
+    }
+  }
+
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
+  @ValueSource(strings = Array("zk", "kraft"))
+  @Timeout(30)
+  def testFollowerCompleteDelayedPurgatoryOnReplication(quorum: String): Unit = {
+    TestUtils.generateAndProduceMessages(brokers, topic, nMessages)
+    // set fetch.max.wait.ms to a value (45 seconds) greater than the timeout (30 seconds) to ensure that the
+    // test only passes when the delayed fetch purgatory is completed after successfully replicating from the leader.
+
+    val totalMessages = initialMessages + nMessages
+    val topicPartition = new TopicPartition(topic, 0)
+    val offsetMap = Map[TopicPartition, Long](
+      topicPartition -> (totalMessages - 1)
```

Comment:
the test fails when the offset is set to `totalMessages`. i first assumed because it was due to delayed HWM propagation but looking at the response from `totalMessages - 1` shows
```
PartitionData(partitionIndex=0, errorCode=0, highWatermark=200, lastStableOffset=200, logStartOffset=0
```
so it does have HWM set to 200. why can't we fetch offset 200?

Reply:
I suppose that the request stays in the purgatory when you fetch offset 200 because there is not record at offset 200 yet and your don't produce anything after the fetch request is sent out.

I think that you test does not really test what you want here. In my mind, you should do the following:
* create the topic with no records.
* fetch at offset 0 with a high max wait. the request should be placed in the purgatory.
* produce one record.
* fetch request should be completed.

Reply:
the suggestion makes sense. i think the point you're getting at is we need to have the fetch sit in purgatory before replication. and i confirmed your suggestion works locally.

however, i'm still unable to grasp why the current logic fails. let's say "there is no record at offset 200 yet" which means replication is not complete. the fetch request will sit in the purgatory and wake up once the record is replicated.

could it be when the record at offset 200 is replicated, the follower's HWM is not updated yet to 200 so the fetch request sits in purgatory and never wakes up because there are no new records?

Reply:
ok, looking deeper into this, here's my understanding as to why fetching offset 200 did not work:

The log end offset points to the next offset the log will append a record, so technically it doesn't exist yet. furthermore, the high watermark is updated to the log end offset. in `Partition.maybeIncrementLeaderHW()`:
```
var newHighWatermark = leaderLogEndOffset
...
leaderLog.maybeIncrementHighWatermark(newHighWatermark)
```
even if the fetch offset is equal to the high watermark, we cannot fetch it if the high watermark is pointing to the log end offset. 

this is done in DelayedFetch where we skip accumulating bytes `if (endOffset.messageOffset (LOE) == fetchOffset.messageOffset (HWM))`

@dajac does that align with your understanding?

Reply:
also, i confirmed that the existing test fails when the follower is already up-to-date. the test above should work generally because the fetch request reaches the follower before the follower is fully replicated. will update the test